cachito/workers/pkg_managers/yarn.py

```diff
@@ -55,3 +64,293 @@ def get_yarn_proxy_repo_username(request_id):
+    nexus_replacements = {}
+
+    for dep_identifier, dep_data in yarn_lock.items():
+        package = pyarn.lockfile.Package.from_dict(dep_identifier, dep_data)
+
+        if package.url:
+            source = package.url
+        elif package.relpath:
+            source = f"file:{package.relpath}"
+        else:
+            raise CachitoError(f"The dependency {package.name}@{package.version} has no source")
+
+        nexus_replacement = None
+
+        non_registry = not package.url or not _is_from_npm_registry(package.url)
+        if non_registry:
+            log.info("The dependency %r is not from the npm registry", f"{package.name}@{source}")
+            # If the file dependency is in the allowlist, do not convert it to a Nexus
+            # dependency, simply ignore it (it will also be ignored by the code that downloads
+            # dependencies later)
+            if package.relpath and normpath(package.relpath) in file_deps_allowlist:
+                log.info("The dependency %r is an allowed exception", package.name)
+            else:
+                nexus_replacement = _convert_to_nexus_hosted(package.name, source, dep_data)
+
+        dep = {
+            "bundled": False,  # yarn.lock does not seem to contain bundled deps at all
+            "dev": False,  # yarn.lock does not state whether a dependency is dev
+            # TODO: ^ this can be solved by cross-referencing with package.json
+            "name": package.name,
+            "version_in_nexus": nexus_replacement["version"] if nexus_replacement else None,
+            "type": "yarn",
+            "version": package.version if not non_registry else source,
+        }
+        deps.append(dep)
+
+        if nexus_replacement:
+            nexus_replacements[dep_identifier] = nexus_replacement
+
+    return deps, nexus_replacements
+
+
+def _is_from_npm_registry(pkg_url):
+    """
+    Check if package is from the NPM registry (which is also the Yarn registry).
+
+    :param str pkg_url: url of the package, in yarn.lock this is always the "resolved" key
+    :rtype: bool
+    """
+    return urlparse(pkg_url).hostname in NPM_REGISTRY_CNAMES
+
+
+def _convert_to_nexus_hosted(dep_name, dep_source, dep_info):
+    """
+    Convert the input dependency not from the NPM registry to a Nexus hosted dependency.
+
+    :param str dep_name: the name of the dependency
+    :param str dep_source: the source (url or relative path) of the dependency
+    :param dict dep_info: the dependency info from the yarn lock file
+    :return: the dependency information of the Nexus hosted version to use in the yarn lock file
+        instead of the original
+    :raise CachitoError: if the dependency is from an unsupported location or has an unexpected
+        format in the lock file
+    """
+    integrity = dep_info.get("integrity")
+    # TODO: integrity may be a space-separated list of integrity values,
+    #   we ignore the possibility for npm, safe to ignore for yarn as well?
```

Comment:
In what cases would yarn v1 generate a separated list here?

Reply:
`yarn install --update-checksums`

Here's how to reproduce:
```shell
cd /tmp
git clone https://github.com/chmeliik/cachito-npm-test.git
cd cachito-npm-test
git checkout yarn-test
podman run --rm -ti -v "$PWD:$PWD:z" -w "$PWD" fedora:33
# [inside the container]
curl --silent --location https://dl.yarnpkg.com/rpm/yarn.repo | sudo tee /etc/yum.repos.d/yarn.repo
dnf -y install yarn git-core
yarn install --update-checksums
git diff
```
Here's an excerpt from the diff:
```diff
diff --git a/yarn.lock b/yarn.lock
index 84b5502..012305f 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -5,12 +5,12 @@
 assertion-error@^1.1.0:
   version "1.1.0"
   resolved "https://registry.yarnpkg.com/assertion-error/-/assertion-error-1.1.0.tgz#e60b6b0e8f301bd97e5375215bda406c85118c0b"
-  integrity sha512-jgsaNduz+ndvGyFt3uSuWqvy4lCnIJiovtouQN5JZHOKCS2QuhEdbcQHFhVksz2N2U9hXJo8odG7ETyWlEeuDw==
+  integrity "sha1-5gtrDo8wG9l+U3UhW9pAbIURjAs= sha512-jgsaNduz+ndvGyFt3uSuWqvy4lCnIJiovtouQN5JZHOKCS2QuhEdbcQHFhVksz2N2U9hXJo8odG7ETyWlEeuDw=="
```

Reply:
It's been quite some time since I looked into this, but I *believe* what yarn does is that it picks the strongest hash and checks only that one

Reply:
Can we do the same, or at least get only one of them when this is the case so our chcks will not break in the future?

Nice catch by the way. I was not aware of this one!

Reply:
So, I tried to figure out what exactly yarn does. It lead me from [here](https://github.com/yarnpkg/yarn/blob/3119382885ea373d3c13d6a846de743eca8c914b/src/fetchers/tarball-fetcher.js#L349) me to a [poorly documented class](https://github.com/npm/ssri#integrity-stream) of the `ssri` package, and my understanding of the yarn code does not match the observed behaviour anyway.

I'll just make Cachito pick the strongest algorithm, if this fails for users they will probably know better what behaviour they expect from yarn.

Fun fact: for git dependencies, yarn does not seem to check integrity at all. Note the comment: https://github.com/yarnpkg/yarn/blob/3119382885ea373d3c13d6a846de743eca8c914b/src/fetchers/git-fetcher.js#L98

:headdesk:

Reply:
done